src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/ApplicationIdRetriever.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2022-2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.bigquery.dwhassessment.hooks.logger;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Optional;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.llap.registry.impl.LlapRegistryService;
+import org.apache.hadoop.hive.ql.MapRedStats;
+import org.apache.hadoop.hive.ql.session.SessionState;
+import org.apache.hadoop.mapred.JobID;
+import org.apache.hadoop.mapreduce.TypeConverter;
+import org.apache.hadoop.yarn.api.records.ApplicationId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Helper for retrieving YARN Application ID depending on the execution mode. */
+public class ApplicationIdRetriever {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ApplicationIdRetriever.class);
+
+  public static Optional<ApplicationId> determineApplicationId(
+      HiveConf conf, ExecutionMode executionMode) {
+    switch (executionMode) {
+      case MR:
+        return determineMapReduceApplicationId();
+      case TEZ:
+        return determineTezApplicationId();
+      case LLAP:
+        return determineLlapApplicationId(conf, executionMode);
+      default:
+        return Optional.empty();
+    }
+  }
+
+  /**
+   * Retrieves Application ID for the Tez application. Application can be reused, but one
+   * application always have only one queue – if queue changes in the session, new application is
+   * created.
+   */
+  private static Optional<ApplicationId> determineTezApplicationId() {
+    try {
+      return Optional.of(
+          SessionState.get().getTezSession().getSession().getAppMasterApplicationId());
+    } catch (NullPointerException e) {
```

Comment:
Assuming the only NPE can happen in this method only (i.e. not inside any of the methods called, like `getTezSession()`), then it can be replaced with:

```
return Optional.ofNullable(Session.get())
  .flatMap(sessionState -> Optional.ofNullable(sessionState.getTezSession()))
  .flatMap(tezSession -> Optional.ofNullable(tezSession.getSession()))
  .flatMap(session -> Optional.ofNullable(session.getAppMasterApplicationId()))
  .or(() -> {
    LOG.info("Failed to retrieve Application ID from Tez session");
    return Optional.empty();
  });
```

Maybe you already know that some of these instances cannot be null, so the corresponding flatMaps could be squashed.

Reply:
Everything after `SessionState.get()` can be null. Maybe even `SessionState.get()` too, but never encountered it.

`.or` might not be present in Java 8. Without it, the code becomes much less nice, so I would leave it as is.